perl5/man/man3/Moose.3
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.15)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "Moose 3"
.TH Moose 3 "2011-01-04" "perl v5.10.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
Moose \- A postmodern object system for Perl 5
.SH "VERSION"
.IX Header "VERSION"
version 1.9902
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 2
\&  package Point;
\&  use Moose; # automatically turns on strict and warnings
\&
\&  has \*(Aqx\*(Aq => (is => \*(Aqrw\*(Aq, isa => \*(AqInt\*(Aq);
\&  has \*(Aqy\*(Aq => (is => \*(Aqrw\*(Aq, isa => \*(AqInt\*(Aq);
\&
\&  sub clear {
\&      my $self = shift;
\&      $self\->x(0);
\&      $self\->y(0);
\&  }
\&
\&  package Point3D;
\&  use Moose;
\&
\&  extends \*(AqPoint\*(Aq;
\&
\&  has \*(Aqz\*(Aq => (is => \*(Aqrw\*(Aq, isa => \*(AqInt\*(Aq);
\&
\&  after \*(Aqclear\*(Aq => sub {
\&      my $self = shift;
\&      $self\->z(0);
\&  };
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Moose is an extension of the Perl 5 object system.
.PP
The main goal of Moose is to make Perl 5 Object Oriented programming
easier, more consistent, and less tedious. With Moose you can think
more about what you want to do and less about the mechanics of \s-1OOP\s0.
.PP
Additionally, Moose is built on top of Class::MOP, which is a
metaclass system for Perl 5. This means that Moose not only makes
building normal Perl 5 objects better, but it provides the power of
metaclass programming as well.
.SS "New to Moose?"
.IX Subsection "New to Moose?"
If you're new to Moose, the best place to start is the
Moose::Manual docs, followed by the Moose::Cookbook. The intro
will show you what Moose is, and how it makes Perl 5 \s-1OO\s0 better.
.PP
The cookbook recipes on Moose basics will get you up to speed with
many of Moose's features quickly. Once you have an idea of what Moose
can do, you can use the \s-1API\s0 documentation to get more detail on
features which interest you.
.SS "Moose Extensions"
.IX Subsection "Moose Extensions"
The \f(CW\*(C`MooseX::\*(C'\fR namespace is the official place to find Moose extensions.
These extensions can be found on the \s-1CPAN\s0.  The easiest way to find them
is to search for them (<http://search.cpan.org/search?query=MooseX::>),
or to examine Task::Moose which aims to keep an up-to-date, easily
installable list of Moose extensions.
.SH "TRANSLATIONS"
.IX Header "TRANSLATIONS"
Much of the Moose documentation has been translated into other languages.
.IP "Japanese" 4
.IX Item "Japanese"
Japanese docs can be found at
http://perldoc.perlassociation.org/pod/Moose\-Doc\-JA/index.html <http://perldoc.perlassociation.org/pod/Moose-Doc-JA/index.html>. The
source \s-1POD\s0 files can be found in GitHub:
http://github.com/jpa/Moose\-Doc\-JA <http://github.com/jpa/Moose-Doc-JA>
.SH "BUILDING CLASSES WITH MOOSE"
.IX Header "BUILDING CLASSES WITH MOOSE"
Moose makes every attempt to provide as much convenience as possible during
class construction/definition, but still stay out of your way if you want it
to. Here are a few items to note when building classes with Moose.
.PP
When you \f(CW\*(C`use Moose\*(C'\fR, Moose will set the class's parent class to
Moose::Object, \fIunless\fR the class using Moose already has a parent
class. In addition, specifying a parent with \f(CW\*(C`extends\*(C'\fR will change the parent
class.
.PP
Moose will also manage all attributes (including inherited ones) that are
defined with \f(CW\*(C`has\*(C'\fR. And (assuming you call \f(CW\*(C`new\*(C'\fR, which is inherited from
Moose::Object) this includes properly initializing all instance slots,
setting defaults where appropriate, and performing any type constraint checking
or coercion.
.SH "PROVIDED METHODS"
.IX Header "PROVIDED METHODS"
Moose provides a number of methods to all your classes, mostly through the
inheritance of Moose::Object. There is however, one exception.
.IP "\fBmeta\fR" 4
.IX Item "meta"
This is a method which provides access to the current class's metaclass.
.SH "EXPORTED FUNCTIONS"
.IX Header "EXPORTED FUNCTIONS"
Moose will export a number of functions into the class's namespace which
may then be used to set up the class. These functions all work directly
on the current class.
.IP "\fBextends (@superclasses)\fR" 4
.IX Item "extends (@superclasses)"
This function will set the superclass(es) for the current class.
.Sp
This approach is recommended instead of \f(CW\*(C`use base\*(C'\fR, because \f(CW\*(C`use base\*(C'\fR
actually \f(CW\*(C`push\*(C'\fRes onto the class's \f(CW@ISA\fR, whereas \f(CW\*(C`extends\*(C'\fR will
replace it. This is important to ensure that classes which do not have
superclasses still properly inherit from Moose::Object.
.Sp
Each superclass can be followed by a hash reference with options. Currently,
only \-version is recognized:
.Sp
.Vb 2
\&    extends \*(AqMy::Parent\*(Aq      => { \-version => 0.01 },
\&            \*(AqMy::OtherParent\*(Aq => { \-version => 0.03 };
.Ve
.Sp
An exception will be thrown if the version requirements are not
satisfied.
.IP "\fBwith (@roles)\fR" 4
.IX Item "with (@roles)"
This will apply a given set of \f(CW@roles\fR to the local class.
.Sp
Like with \f(CW\*(C`extends\*(C'\fR, each specified role can be followed by a hash
reference with a \-version option:
.Sp
.Vb 2
\&    with \*(AqMy::Role\*(Aq      => { \-version => 0.32 },
\&         \*(AqMy::Otherrole\*(Aq => { \-version => 0.23 };
.Ve
.Sp
The specified version requirements must be satisfied, otherwise an
exception will be thrown.
.Sp
If your role takes options or arguments, they can be passed along in the
hash reference as well.
.ie n .IP "\fBhas \fB$name\fB|@$names => \f(BI%options\fB\fR" 4
.el .IP "\fBhas \f(CB$name\fB|@$names => \f(CB%options\fB\fR" 4
.IX Item "has $name|@$names => %options"
This will install an attribute of a given \f(CW$name\fR into the current class. If
the first parameter is an array reference, it will create an attribute for
every \f(CW$name\fR in the list. The \f(CW%options\fR are the same as those provided by
Class::MOP::Attribute, in addition to the list below which are provided by
Moose (Moose::Meta::Attribute to be more specific):
.RS 4
.IP "\fIis => 'rw'|'ro'\fR" 4
.IX Item "is => 'rw'|'ro'"
The \fIis\fR option accepts either \fIrw\fR (for read/write) or \fIro\fR (for read
only). These will create either a read/write accessor or a read-only
accessor respectively, using the same name as the \f(CW$name\fR of the attribute.
.Sp
If you need more control over how your accessors are named, you can
use the reader,
writer and
accessor options inherited from
Class::MOP::Attribute, however if you use those, you won't need the
\&\fIis\fR option.
.ie n .IP "\fIisa => \fI$type_name\fI\fR" 4
.el .IP "\fIisa => \f(CI$type_name\fI\fR" 4
.IX Item "isa => $type_name"
The \fIisa\fR option uses Moose's type constraint facilities to set up runtime
type checking for this attribute. Moose will perform the checks during class
construction, and within any accessors. The \f(CW$type_name\fR argument must be a
string. The string may be either a class name or a type defined using
Moose's type definition features. (Refer to Moose::Util::TypeConstraints
for information on how to define a new type, and how to retrieve type meta-data).
.IP "\fIcoerce => (1|0)\fR" 4
.IX Item "coerce => (1|0)"
This will attempt to use coercion with the supplied type constraint to change
the value passed into any accessors or constructors. You \fBmust\fR supply a type
constraint, and that type constraint \fBmust\fR define a coercion. See
Moose::Cookbook::Basics::Recipe5 for an example.
.ie n .IP "\fIdoes => \fI$role_name\fI\fR" 4
.el .IP "\fIdoes => \f(CI$role_name\fI\fR" 4
.IX Item "does => $role_name"
This will accept the name of a role which the value stored in this attribute
is expected to have consumed.
.IP "\fIrequired => (1|0)\fR" 4
.IX Item "required => (1|0)"
This marks the attribute as being required. This means a value must be
supplied during class construction, \fIor\fR the attribute must be lazy
and have either a default or a builder. Note that c<required> does not
say anything about the attribute's value, which can be \f(CW\*(C`undef\*(C'\fR.
.IP "\fIweak_ref => (1|0)\fR" 4
.IX Item "weak_ref => (1|0)"
This will tell the class to store the value of this attribute as a weakened
reference. If an attribute is a weakened reference, it \fBcannot\fR also be
coerced.
.IP "\fIlazy => (1|0)\fR" 4
.IX Item "lazy => (1|0)"
This will tell the class to not create this slot until absolutely necessary.
If an attribute is marked as lazy it \fBmust\fR have a default supplied.
.IP "\fIauto_deref => (1|0)\fR" 4
.IX Item "auto_deref => (1|0)"
This tells the accessor to automatically dereference the value of this
attribute when called in list context.  The accessor will still return a
reference when called in scalar context.  If this behavior isn't desirable,
\&\*(L"elements\*(R" in Moose::Meta::Attribute::Native::Trait::Array or
\&\*(L"elements\*(R" in Moose::Meta::Attribute::Native::Trait::Hash may be a better
choice.  The \fIauto_deref\fR option is only legal if your \fIisa\fR option is
either \f(CW\*(C`ArrayRef\*(C'\fR or \f(CW\*(C`HashRef\*(C'\fR.
.ie n .IP "\fItrigger => \fI$code\fI\fR" 4
.el .IP "\fItrigger => \f(CI$code\fI\fR" 4
.IX Item "trigger => $code"
The \fItrigger\fR option is a \s-1CODE\s0 reference which will be called after
the value of the attribute is set. The \s-1CODE\s0 ref is passed the
instance itself, the updated value, and the original value if the
attribute was already set.
.Sp
You \fBcan\fR have a trigger on a read-only attribute.
.Sp
\&\fB\s-1NOTE:\s0\fR Triggers will only fire when you \fBassign\fR to the attribute,
either in the constructor, or using the writer. Default and built values will
\&\fBnot\fR cause the trigger to be fired.
.IP "\fIhandles => \s-1ARRAY\s0 | \s-1HASH\s0 | \s-1REGEXP\s0 | \s-1ROLE\s0 | \s-1ROLETYPE\s0 | \s-1DUCKTYPE\s0 | \s-1CODE\s0\fR" 4
.IX Item "handles => ARRAY | HASH | REGEXP | ROLE | ROLETYPE | DUCKTYPE | CODE"
The \fIhandles\fR option provides Moose classes with automated delegation features.
This is a pretty complex and powerful option. It accepts many different option
formats, each with its own benefits and drawbacks.
.Sp
\&\fB\s-1NOTE:\s0\fR The class being delegated to does not need to be a Moose based class,
which is why this feature is especially useful when wrapping non-Moose classes.
.Sp
All \fIhandles\fR option formats share the following traits:
.Sp
You cannot override a locally defined method with a delegated method; an
exception will be thrown if you try. That is to say, if you define \f(CW\*(C`foo\*(C'\fR in
your class, you cannot override it with a delegated \f(CW\*(C`foo\*(C'\fR. This is almost never
something you would want to do, and if it is, you should do it by hand and not
use Moose.
.Sp
You cannot override any of the methods found in Moose::Object, or the \f(CW\*(C`BUILD\*(C'\fR
and \f(CW\*(C`DEMOLISH\*(C'\fR methods. These will not throw an exception, but will silently
move on to the next method in the list. My reasoning for this is that you would
almost never want to do this, since it usually breaks your class. As with
overriding locally defined methods, if you do want to do this, you should do it
manually, not with Moose.
.Sp
You do not \fIneed\fR to have a reader (or accessor) for the attribute in order
to delegate to it. Moose will create a means of accessing the value for you,
however this will be several times \fBless\fR efficient then if you had given
the attribute a reader (or accessor) to use.
.Sp
Below is the documentation for each option format:
.RS 4
.ie n .IP """ARRAY""" 4
.el .IP "\f(CWARRAY\fR" 4
.IX Item "ARRAY"
This is the most common usage for \fIhandles\fR. You basically pass a list of
method names to be delegated, and Moose will install a delegation method
for each one.
.ie n .IP """HASH""" 4
.el .IP "\f(CWHASH\fR" 4
.IX Item "HASH"
This is the second most common usage for \fIhandles\fR. Instead of a list of
method names, you pass a \s-1HASH\s0 ref where each key is the method name you
want installed locally, and its value is the name of the original method
in the class being delegated to.
.Sp
This can be very useful for recursive classes like trees. Here is a
quick example (soon to be expanded into a Moose::Cookbook recipe):
.Sp
.Vb 2
\&  package Tree;
\&  use Moose;
\&
\&  has \*(Aqnode\*(Aq => (is => \*(Aqrw\*(Aq, isa => \*(AqAny\*(Aq);
\&
\&  has \*(Aqchildren\*(Aq => (
\&      is      => \*(Aqro\*(Aq,
\&      isa     => \*(AqArrayRef\*(Aq,
\&      default => sub { [] }
\&  );
\&
\&  has \*(Aqparent\*(Aq => (
\&      is          => \*(Aqrw\*(Aq,
\&      isa         => \*(AqTree\*(Aq,
\&      weak_ref    => 1,
\&      handles     => {
\&          parent_node => \*(Aqnode\*(Aq,
\&          siblings    => \*(Aqchildren\*(Aq,
\&      }
\&  );
.Ve
.Sp
In this example, the Tree package gets \f(CW\*(C`parent_node\*(C'\fR and \f(CW\*(C`siblings\*(C'\fR methods,
which delegate to the \f(CW\*(C`node\*(C'\fR and \f(CW\*(C`children\*(C'\fR methods (respectively) of the Tree
instance stored in the \f(CW\*(C`parent\*(C'\fR slot.
.Sp
You may also use an array reference to curry arguments to the original method.
.Sp
.Vb 4
\&  has \*(Aqthing\*(Aq => (
\&      ...
\&      handles => { set_foo => [ set => \*(Aqfoo\*(Aq ] },
\&  );
\&
\&  # $self\->set_foo(...) calls $self\->thing\->set(\*(Aqfoo\*(Aq, ...)
.Ve
.Sp
The first element of the array reference is the original method name, and the
rest is a list of curried arguments.
.ie n .IP """REGEXP""" 4
.el .IP "\f(CWREGEXP\fR" 4
.IX Item "REGEXP"
The regexp option works very similar to the \s-1ARRAY\s0 option, except that it builds
the list of methods for you. It starts by collecting all possible methods of the
class being delegated to, then filters that list using the regexp supplied here.
.Sp
\&\fB\s-1NOTE:\s0\fR An \fIisa\fR option is required when using the regexp option format. This
is so that we can determine (at compile time) the method list from the class.
Without an \fIisa\fR this is just not possible.
.ie n .IP """ROLE"" or ""ROLETYPE""" 4
.el .IP "\f(CWROLE\fR or \f(CWROLETYPE\fR" 4
.IX Item "ROLE or ROLETYPE"
With the role option, you specify the name of a role or a
role type whose \*(L"interface\*(R" then becomes
the list of methods to handle. The \*(L"interface\*(R" can be defined as; the methods
of the role and any required methods of the role. It should be noted that this
does \fBnot\fR include any method modifiers or generated attribute methods (which
is consistent with role composition).
.ie n .IP """DUCKTYPE""" 4
.el .IP "\f(CWDUCKTYPE\fR" 4
.IX Item "DUCKTYPE"
With the duck type option, you pass a duck type object whose \*(L"interface\*(R" then
becomes the list of methods to handle. The \*(L"interface\*(R" can be defined as the
list of methods passed to \f(CW\*(C`duck_type\*(C'\fR to create a duck type object. For more
information on \f(CW\*(C`duck_type\*(C'\fR please check
Moose::Util::TypeConstraints.
.ie n .IP """CODE""" 4
.el .IP "\f(CWCODE\fR" 4
.IX Item "CODE"
This is the option to use when you really want to do something funky. You should
only use it if you really know what you are doing, as it involves manual
metaclass twiddling.
.Sp
This takes a code reference, which should expect two arguments. The first is the
attribute meta-object this \fIhandles\fR is attached to. The second is the
metaclass of the class being delegated to. It expects you to return a hash (not
a \s-1HASH\s0 ref) of the methods you want mapped.
.RE
.RS 4
.RE
.ie n .IP "\fImetaclass => \fI$metaclass_name\fI\fR" 4
.el .IP "\fImetaclass => \f(CI$metaclass_name\fI\fR" 4
.IX Item "metaclass => $metaclass_name"
This tells the class to use a custom attribute metaclass for this particular
attribute. Custom attribute metaclasses are useful for extending the
capabilities of the \fIhas\fR keyword: they are the simplest way to extend the \s-1MOP\s0,
but they are still a fairly advanced topic and too much to cover here. See
Moose::Cookbook::Meta::Recipe1 for more information.
.Sp
See \*(L"Metaclass and Trait Name Resolution\*(R" for details on how a metaclass name
is resolved to a class name.
.ie n .IP "\fItraits => [ \fI@role_names\fI ]\fR" 4
.el .IP "\fItraits => [ \f(CI@role_names\fI ]\fR" 4
.IX Item "traits => [ @role_names ]"
This tells Moose to take the list of \f(CW@role_names\fR and apply them to the
attribute meta-object. This is very similar to the \fImetaclass\fR option, but
allows you to use more than one extension at a time.
.Sp
See \*(L"Metaclass and Trait Name Resolution\*(R" for details on how a trait name is
resolved to a role name.
.Sp
Also see Moose::Cookbook::Meta::Recipe3 for a metaclass trait
example.
.IP "\fIbuilder\fR => Str" 4
.IX Item "builder => Str"
The value of this key is the name of the method that will be called to
obtain the value used to initialize the attribute. See the builder
option docs in Class::MOP::Attribute
and/or Moose::Cookbook::Basics::Recipe8 for more information.
.IP "\fIdefault\fR => \s-1SCALAR\s0 | \s-1CODE\s0" 4
.IX Item "default => SCALAR | CODE"
The value of this key is the default value which will initialize the attribute.
.Sp
\&\s-1NOTE:\s0 If the value is a simple scalar (string or number), then it can
be just passed as is.  However, if you wish to initialize it with a
\&\s-1HASH\s0 or \s-1ARRAY\s0 ref, then you need to wrap that inside a \s-1CODE\s0 reference.
See the default option docs in
Class::MOP::Attribute for more
information.
.IP "\fIclearer\fR => Str" 4
.IX Item "clearer => Str"
Creates a method allowing you to clear the value. See the clearer option
docs in Class::MOP::Attribute for more
information.
.IP "\fIpredicate\fR => Str" 4
.IX Item "predicate => Str"
Creates a method to perform a basic test to see if a value has been set in the
attribute. See the predicate option docs in
Class::MOP::Attribute for more information.
.IP "\fIlazy_build\fR => (0|1)" 4
.IX Item "lazy_build => (0|1)"
Automatically define lazy => 1 as well as builder => \*(L"_build_$attr\*(R", clearer =>
"clear_$attr', predicate => 'has_$attr' unless they are already defined.
.IP "\fIinitializer\fR => Str" 4
.IX Item "initializer => Str"
This may be a method name (referring to a method on the class with
this attribute) or a \s-1CODE\s0 ref.  The initializer is used to set the
attribute value on an instance when the attribute is set during
instance initialization (but not when the value is being assigned
to). See the initializer option docs in
Class::MOP::Attribute for more
information.
.ie n .IP "\fIdocumentation\fR => $string" 4
.el .IP "\fIdocumentation\fR => \f(CW$string\fR" 4
.IX Item "documentation => $string"
An arbitrary string that can be retrieved later by calling \f(CW\*(C`$attr\->documentation\*(C'\fR.
.RE
.RS 4
.RE
.ie n .IP "\fBhas +$name => \fB%options\fB\fR" 4
.el .IP "\fBhas +$name => \f(CB%options\fB\fR" 4
.IX Item "has +$name => %options"
This is variation on the normal attribute creator \f(CW\*(C`has\*(C'\fR which allows you to
clone and extend an attribute from a superclass or from a role. Here is an
example of the superclass usage:
.Sp
.Vb 2
\&  package Foo;
\&  use Moose;
\&
\&  has \*(Aqmessage\*(Aq => (
\&      is      => \*(Aqrw\*(Aq,
\&      isa     => \*(AqStr\*(Aq,
\&      default => \*(AqHello, I am a Foo\*(Aq
\&  );
\&
\&  package My::Foo;
\&  use Moose;
\&
\&  extends \*(AqFoo\*(Aq;
\&
\&  has \*(Aq+message\*(Aq => (default => \*(AqHello I am My::Foo\*(Aq);
.Ve
.Sp
What is happening here is that \fBMy::Foo\fR is cloning the \f(CW\*(C`message\*(C'\fR attribute
from its parent class \fBFoo\fR, retaining the \f(CW\*(C`is => \*(Aqrw\*(Aq\*(C'\fR and \f(CW\*(C`isa =>
\&\*(AqStr\*(Aq\*(C'\fR characteristics, but changing the value in \f(CW\*(C`default\*(C'\fR.
.Sp
Here is another example, but within the context of a role:
.Sp
.Vb 2
\&  package Foo::Role;
\&  use Moose::Role;
\&
\&  has \*(Aqmessage\*(Aq => (
\&      is      => \*(Aqrw\*(Aq,
\&      isa     => \*(AqStr\*(Aq,
\&      default => \*(AqHello, I am a Foo\*(Aq
\&  );
\&
\&  package My::Foo;
\&  use Moose;
\&
\&  with \*(AqFoo::Role\*(Aq;
\&
\&  has \*(Aq+message\*(Aq => (default => \*(AqHello I am My::Foo\*(Aq);
.Ve
.Sp
In this case, we are basically taking the attribute which the role supplied
and altering it within the bounds of this feature.
.Sp
Note that you can only extend an attribute from either a superclass or a role,
you cannot extend an attribute in a role that composes over an attribute from
another role.
.Sp
Aside from where the attributes come from (one from superclass, the other
from a role), this feature works exactly the same. This feature is restricted
somewhat, so as to try and force at least \fIsome\fR sanity into it. You are only
allowed to change the following attributes:
.RS 4
.IP "\fIdefault\fR" 4
.IX Item "default"
Change the default value of an attribute.
.IP "\fIcoerce\fR" 4
.IX Item "coerce"
Change whether the attribute attempts to coerce a value passed to it.
.IP "\fIrequired\fR" 4
.IX Item "required"
Change if the attribute is required to have a value.
.IP "\fIdocumentation\fR" 4
.IX Item "documentation"
Change the documentation string associated with the attribute.
.IP "\fIlazy\fR" 4
.IX Item "lazy"
Change if the attribute lazily initializes the slot.
.IP "\fIisa\fR" 4
.IX Item "isa"
You \fIare\fR allowed to change the type without restriction.
.Sp
It is recommended that you use this freedom with caution. We used to
only allow for extension only if the type was a subtype of the parent's
type, but we felt that was too restrictive and is better left as a
policy decision.
.IP "\fIhandles\fR" 4
.IX Item "handles"
You are allowed to \fBadd\fR a new \f(CW\*(C`handles\*(C'\fR definition, but you are \fBnot\fR
allowed to \fIchange\fR one.
.IP "\fIbuilder\fR" 4
.IX Item "builder"
You are allowed to \fBadd\fR a new \f(CW\*(C`builder\*(C'\fR definition, but you are \fBnot\fR
allowed to \fIchange\fR one.
.IP "\fImetaclass\fR" 4
.IX Item "metaclass"
You are allowed to \fBadd\fR a new \f(CW\*(C`metaclass\*(C'\fR definition, but you are
\&\fBnot\fR allowed to \fIchange\fR one.
.IP "\fItraits\fR" 4
.IX Item "traits"
You are allowed to \fBadd\fR additional traits to the \f(CW\*(C`traits\*(C'\fR definition.
These traits will be composed into the attribute, but preexisting traits
\&\fBare not\fR overridden, or removed.
.RE
.RS 4
.RE
.ie n .IP "\fBbefore \fB$name\fB|@names|\e@names|qr/.../ => sub { ... }\fR" 4
.el .IP "\fBbefore \f(CB$name\fB|@names|\e@names|qr/.../ => sub { ... }\fR" 4
.IX Item "before $name|@names|@names|qr/.../ => sub { ... }"
.PD 0
.ie n .IP "\fBafter \fB$name\fB|@names|\e@names|qr/.../ => sub { ... }\fR" 4
.el .IP "\fBafter \f(CB$name\fB|@names|\e@names|qr/.../ => sub { ... }\fR" 4
.IX Item "after $name|@names|@names|qr/.../ => sub { ... }"
.ie n .IP "\fBaround \fB$name\fB|@names|\e@names|qr/.../ => sub { ... }\fR" 4
.el .IP "\fBaround \f(CB$name\fB|@names|\e@names|qr/.../ => sub { ... }\fR" 4
.IX Item "around $name|@names|@names|qr/.../ => sub { ... }"
.PD
These three items are syntactic sugar for the before, after, and around method
modifier features that Class::MOP provides. More information on these may be
found in Moose::Manual::MethodModifiers and the
Class::MOP::Class documentation.
.IP "\fBsuper\fR" 4
.IX Item "super"
The keyword \f(CW\*(C`super\*(C'\fR is a no-op when called outside of an \f(CW\*(C`override\*(C'\fR method. In
the context of an \f(CW\*(C`override\*(C'\fR method, it will call the next most appropriate
superclass method with the same arguments as the original method.
.IP "\fBoverride ($name, &sub)\fR" 4
.IX Item "override ($name, &sub)"

Changes
1.100990 - 2010-04-09
  - initial release

lib/MooseX/Attribute/Deflator/Moose.pm
#
# This file is part of MooseX-Attribute-Deflator
#
# This software is Copyright (c) 2011 by Moritz Onken.
#
# This is free software, licensed under:
#
#   The (three-clause) BSD License
#
package MooseX::Attribute::Deflator::Moose;
BEGIN {
  $MooseX::Attribute::Deflator::Moose::VERSION = '1.130002';
}
# ABSTRACT: Deflators for Moose type constraints

use MooseX::Attribute::Deflator;
use JSON;


deflate 'HashRef', via { JSON::encode_json($_) };
inflate 'HashRef', via { JSON::decode_json($_) };

deflate 'ArrayRef', via { JSON::encode_json($_) };
inflate 'ArrayRef', via { JSON::decode_json($_) };

deflate 'ScalarRef', via { $$_ };
inflate 'ScalarRef', via { \$_ };

deflate 'Item', via { $_ };
inflate 'Item', via { $_ };

deflate 'HashRef[]', via {
    my ($attr, $constraint, $deflate) = @_;
    my $value = {%$_};
    while(my($k,$v) = each %$value) {
        $value->{$k} = $deflate->($value->{$k}, $constraint->type_parameter);
    }
    return $deflate->($value, $constraint->parent);
};

inflate 'HashRef[]', via {
    my ($attr, $constraint, $inflate) = @_;
    my $value = $inflate->($_, $constraint->parent);
    while(my($k,$v) = each %$value) {
        $value->{$k} = $inflate->($value->{$k}, $constraint->type_parameter);
    }
    return $value;
};

deflate 'ArrayRef[]', via {
    my ($attr, $constraint, $deflate) = @_;
    my $value = [@$_];
    $_ = $deflate->($_, $constraint->type_parameter) for(@$value);
    return $deflate->($value, $constraint->parent);
};

inflate 'ArrayRef[]', via {
    my ($attr, $constraint, $inflate) = @_;
    my $value = $inflate->($_, $constraint->parent);
    $_ = $inflate->($_, $constraint->type_parameter) for(@$value);
    return $value;
};

deflate 'Maybe[]', via {
    my ($attr, $constraint, $deflate) = @_;
    return $deflate->($_, $constraint->type_parameter);
};

inflate 'Maybe[]', via {
    my ($attr, $constraint, $inflate) = @_;
    return $inflate->($_, $constraint->type_parameter);
};

deflate 'ScalarRef[]', via {
    my ($attr, $constraint, $deflate) = @_;
    return ${$deflate->($_, $constraint->type_parameter)};
};

inflate 'ScalarRef[]', via {
    my ($attr, $constraint, $inflate) = @_;
    return \$inflate->($_, $constraint->type_parameter);
};

1;



=pod

=head1 NAME

MooseX::Attribute::Deflator::Moose - Deflators for Moose type constraints

=head1 VERSION

version 1.130002

=head1 SYNOPSIS

  use MooseX::Attribute::Deflator::Moose;

=head1 DESCRIPTION

Using this module registers sane type deflators and inflators for Moose's built in types.

Some notes:

=over

=item * HashRef and ArrayRef deflate/inflate using JSON

=item * ScalarRef is dereferenced on deflation and returns a reference on inflation

=back

=head1 AUTHOR

Moritz Onken

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2011 by Moritz Onken.

This is free software, licensed under:

  The (three-clause) BSD License

=cut


__END__

